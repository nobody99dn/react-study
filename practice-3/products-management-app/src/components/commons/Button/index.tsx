// Libraries
import React from 'react';

// Constants
import { ButtonVariants } from '@constants/index';

// Style
import './index.css';

interface ButtonProps {
  title: string;
  variant?: ButtonVariants;
  isDisabled?: boolean;
  isLoading?: boolean;
  handleClick?: () => void;
}

const Button: React.FC<ButtonProps> = ({
  title,
  variant = ButtonVariants.Default,
  isDisabled = false,
  isLoading = false,
  handleClick
}) => {
  return (
    <button
      className={`btn btn-${variant} ${
        isDisabled ? 'btn-disabled' : ''
      }`.trim()}
      onClick={handleClick}
      disabled={isDisabled}
    >
      {isLoading ? 'Loading...' : title}
    </button>
  );
};

export default Button;
