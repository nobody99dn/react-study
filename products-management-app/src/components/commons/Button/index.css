.btn {
  border-radius: 3px;
  cursor: pointer;
  font-size: 16px;
  letter-spacing: 0.03rem;
  padding: 0.5rem 1.5rem;
}

.btn:hover,
.disabled {
  opacity: 0.5;
}

/* - Default Button */
/* - ---------------------------------------------- */
.btn-default {
  background-color: var(--white);
  border: 1px solid var(--primary);
  color: var(--dark);
}

/* - Primary Button */
/* - ---------------------------------------------- */
.btn-primary {
  background-color: var(--primary);
  border: 1px solid var(--primary);
  color: var(--white);
}

/* - Secondary Button */
/* - ---------------------------------------------- */
.btn-secondary {
  background-color: var(--danger);
  border: 1px solid var(--danger);
  color: var(--white);
}

/* - Disable Button */
/* - ---------------------------------------------- */
.disabled {
  background-color: var(--light-gray);
  border: 1px solid var(--light-gray);
  cursor: auto;
}
