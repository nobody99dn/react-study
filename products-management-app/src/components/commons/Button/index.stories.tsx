// Libraries
import React, { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Button from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';

// Constants
import { ButtonVariants } from '@constants/types';

export default {
  title: 'Components/Button',
  component: Button
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  title: 'Button'
};

export const Primary = Template.bind({});
Primary.args = {
  title: 'Button',
  variant: ButtonVariants.Primary
};

export const Secondary = Template.bind({});
Secondary.args = {
  title: 'Button',
  variant: ButtonVariants.Secondary
};

export const Disabled = Template.bind({});
Disabled.args = {
  title: 'Button',
  isDisabled: true
};
