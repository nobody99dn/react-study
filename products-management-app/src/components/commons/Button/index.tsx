// Libraries
import React from 'react';

// Constants
import { ButtonVariants } from '@constants/types';

// Style
import './index.css';

interface ButtonProps {
  title: string;
  variant?: ButtonVariants;
  isDisabled?: boolean;
  isLoading?: boolean;
  onClick?: () => void;
}

const Button: React.FC<ButtonProps> = ({
  title,
  variant = ButtonVariants.Default,
  isDisabled = false,
  isLoading = false,
  onClick
}) => {
  return (
    <button
      className={`btn btn-${variant} ${
        isDisabled ? 'btn-disabled' : ''
      }`.trim()}
      onClick={onClick}
      disabled={isDisabled}
    >
      {isLoading ? 'Loading...' : title}
    </button>
  );
};

export default Button;
