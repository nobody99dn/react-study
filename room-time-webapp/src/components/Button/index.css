.btn {
  border: 0px;
  border-radius: 3px;
  cursor: pointer;
  font-size: 15px;
  letter-spacing: 0.03rem;
  padding: 20px 50px;
}

.btn:hover{ 
  opacity: 0.8;
}

.btn-primary {
  background-color: var(--light-white-color);
  color: var(--highlight-color);
}

.btn-secondary {
  background-color: var(--highlight-color);
  color: var(--light-white-color);
}

.btn-outline {
  background: var(--invisible-color);
  border: 2px solid var(--light-gray-color);
  color: var(--light-gray-color);
}

.btn-ghost{ 
  background-color: var(--invisible-color);
  border: 2px solid var(--light-gray-color);
  color: var(--dark-green-color);
}

.btn-ghost:hover {
  border-color: var(--dark-green-color);
}
