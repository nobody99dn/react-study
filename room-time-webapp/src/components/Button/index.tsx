/* eslint-disable @typescript-eslint/no-useless-constructor */

// Libraries
import { ReactNode } from "react";
import React from 'react';

// Style
import './index.css';

interface ButtonProps {
  children: ReactNode;
  theme?: 'primary' | 'secondary' | 'outline' | 'ghost';
  [props: string]: any;
}

export const Button: React.FC<ButtonProps> = ({
  children,
  theme = 'primary',
  ...props
}) => (<button className={`btn btn-${theme}`} {...props}>{children}</button>);

