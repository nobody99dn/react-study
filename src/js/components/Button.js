export const Button = (id = '', className = '', content = 'Button') => `
  <button
    class="${['btn', className].join(' ').trim()}"
    id="${id}"
  >
    ${content}
  </button>
`
