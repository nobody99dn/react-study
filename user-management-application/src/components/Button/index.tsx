import React from "react";
import classes from "./index.module.css";

enum TypeButton {
  primary = "primary",
  secondary = "secondary",
}

interface ButtonProps {
  children?: any;
  icon?: any;
  type?: TypeButton;
}

export const Button: React.FC<ButtonProps> = ({
  icon,
  children,
  type = TypeButton.primary,
}) => {
  return (
    <button
      className={`${classes.btn} ${
        type === TypeButton.primary && classes["btn--primary"]
      } `}
    >
      {icon}
      {children}
    </button>
  );
};
