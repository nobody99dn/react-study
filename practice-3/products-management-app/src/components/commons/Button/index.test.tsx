// Libraries
import { render, act, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';

// Components
import Button from './index';

describe('button component', () => {
  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<Button title='Button' />).toJSON();
    expect(defaultTree).toMatchSnapshot();

    const disableTree = create(<Button title='Disable Button' isDisabled />);
    expect(disableTree).toMatchSnapshot();
  });

  test('should render button with Button', () => {
    act(() => {
      render(<Button title='Button' />, { container });
    });

    expect(container.textContent).toMatch('Button');
  });

  test('should render loading button', () => {
    act(() => {
      render(<Button title='Button' isLoading />, { container });
    });

    const button: HTMLElement = screen.getByRole('button');

    expect(button.textContent).toMatch('Loading...');
  });

  test('should be click 2 times', () => {
    const myMock = jest.fn();
    act(() => {
      render(<Button title='Button' onClick={myMock} />, { container });
    });

    const button: HTMLElement = screen.getByRole('button');

    fireEvent.click(button);
    fireEvent.click(button);

    expect(myMock).toHaveBeenCalledTimes(2);
  });
});
