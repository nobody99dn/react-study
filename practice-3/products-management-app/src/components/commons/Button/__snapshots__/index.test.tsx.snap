// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<button
  className="btn btn-default"
  disabled={false}
>
  Button
</button>
`;

exports[`Button component should render correctly 2`] = `
<button
  className="btn btn-default btn-disabled"
  disabled={true}
>
  Disable Button
</button>
`;

exports[`button component should render correctly 1`] = `
<button
  className="btn btn-default"
  disabled={false}
>
  Button
</button>
`;

exports[`button component should render correctly 2`] = `
<button
  className="btn btn-default btn-disabled"
  disabled={true}
>
  Disable Button
</button>
`;
